### PR TITLE
feat: add eslint no-setter-return rule

### PR DESCRIPTION
**Pull request by anonrig**
Work in progress.

### CODE REVIEW

**Comment by sno2** [comment]
Hello, I looked at implementing this rule before the `no-eval`. However, it gave me a TS lint error and I wasn't sure if we should add double diagnostics. You can see it in the CI:

```
thread 'rules::no_setter_return::test' panicked at '[  × TS1108: A 'return' statement can only be used within a function body
```

**Comment by anonrig** [comment]
> Hello, I looked at implementing this rule before the `no-eval`. However, it gave me a TS lint error and I wasn't sure if we should add double diagnostics. You can see it in the CI:

Yes, I'm seeing a similar error message as well. I'll investigate on it.

File: crates/oxc_linter/src/rules/no_setter_return.rs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+            None,
+        ),
+        (
+            "x = function Object() { Object.defineProperty(foo, 'bar', { set(val) { return 1; } }) }",
+            None,
+        ),
+    ];
+
+    let fail = vec![
+        // ("({ set a(val){ return val + 1; } })", None),
```
Comment (anonrig) [inline crates/oxc_linter/src/rules/no_setter_return.rs:223]: @Boshen Objects having setter functions does not get detected as MethodDefinition::Setter AstNode kind.

**Comment by Boshen** [comment]
> Hello, I looked at implementing this rule before the `no-eval`. However, it gave me a TS lint error and I wasn't sure if we should add double diagnostics. You can see it in the CI:
> 
> ```
> thread 'rules::no_setter_return::test' panicked at '[  × TS1108: A 'return' statement can only be used within a function body
> ```

This is may be due to "return outside of function body", which is related to `allow_return_outside_function` set by the parser.

**Comment by Boshen** [comment]
This is one of the rules where the implementation can differ greatly from ESLint, which will make the implementation really easy.

We need to start looking at the semantics - what's the easiest way to tell if a function is a setter? Can we set that information somewhere?

Yes, on the scope:

https://github.com/Boshen/oxc/blob/985b8f21d93da14ed9e8b5cdcb3cb1e1a2bf8e7f/crates/oxc_semantic/src/scope/mod.rs#L41

Do we have a ast walk some where that can set this information?

Yes, in the binder, 

https://github.com/Boshen/oxc/blob/985b8f21d93da14ed9e8b5cdcb3cb1e1a2bf8e7f/crates/oxc_semantic/src/binder.rs#L92

The only slightly harder part is to go up the ast by using `ancestors` to find `MethodDefinition` find out if it's a contructor / getter / setter.

Then it's really easy inside the linter:

Search `ReturnStatemement` where its scope contains a `SetAccessor`

**Comment by Boshen** [comment]
@anonrig I'm so sorry for such a late reply for this PR, I just found a little bit of time to update this PR, please take a look.

**Pull request merged**